Add Razor LSP client to activate in non-liveshare scenarios and only …

dibarbet opened the pull request:
…ish razor diagnostics.

To resolve the issue of activating the c# lsp for razor in non liveshare scenarios without publishing duplicate diagnostics.

This can be removed as soon as we switch over to LSP for diagnostics in all cases.

dibarbet commented:
cc @NTaylorMullen @jasonmalinowski 
I have not yet tested, just wanted to give an idea of what it might look like

NTaylorMullen commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -236,10 +237,29 @@ private async Task<LanguageServer.Protocol.Diagnostic[]> GetDiagnosticsAsync(Doc
+
+        private LanguageServer.Protocol.Range? GetDiagnosticRange(DiagnosticDataLocation? diagnosticDataLocation, SourceText text)
+        {
+            (var startLine, var endLine) = DiagnosticData.GetLinePositions(diagnosticDataLocation, text, useMapped: true);
```
Out of curiosity why isn't `useMapped: true` the default if it falls back to non-mapped locations in the scenario mapped ones don't exist

dibarbet replied:
It's a good question- I'm not sure what the answer is.  I dug a bit today but couldn't find anything to indicate it shouldn't be the default, but I wanted to avoid possibly changing the behavior in non-lsp scenarios so I introduced the flag.  I'll look a bit more tomorrow.

NTaylorMullen commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,7 +220,7 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+        protected virtual async Task PublishDiagnosticsAsync(Document document)
         {
             var diagnostics = await GetDiagnosticsAsync(document, CancellationToken.None).ConfigureAwait(false);
             var publishDiagnosticsParams = new PublishDiagnosticParams { Diagnostics = diagnostics, Uri = document.GetURI() };
```
This `Uri = document.GetURI()` piece doesn't return the re-mapped document location.  However it does have the correct line location now 😄 

![image](https://user-images.githubusercontent.com/2008729/76656609-3d06a200-652d-11ea-926e-c3ca08e08eab.png)

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:
```diff
@@ -272,6 +272,12 @@ public Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancel
+            {
+                // This is a generated Razor document, and they want diagnostics, so let's report it
+                return true;
+            }
```
@mavasani   I'm not qualified to know if this is ok.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:
```diff
@@ -59,6 +60,22 @@ internal static class ProtocolConversions
+            }
+
+            return new Uri(filePath, UriKind.Absolute);
+        }
```
i hate this entire function.

dibarbet replied:
wiling to take suggestions on how to improve this - it's here because of how razor file paths are normalized to have the same representation on mac/linux/windows.
@NTaylorMullen can clarify more.

NTaylorMullen replied:
First off it's definitely not ideal but there's a need to handle "bad" line pragmas anyhow.

The input into this is from a line pragma and Razor happens generates line pragmas with a leading `/` to keep track of files in a platform independent way. A lot of the file APIs that the runtime uses allow a prefixed `/` and so we prefix all of our paths with it to enable having the same path representation on Windows Mac and Linux.
 
Thinking from your perspective, you all will need to have a flow where you're unable to convert a line pragma into a Uri anyhow. Would it be awful to maintain this attempt small normalization? I ask because it'd be a massive change for our language server to move away from normalizing all our platform specific file paths to look the same.

CyrusNajmabadi replied:
i don't hate the URI.  But i can't stand that we're having to munge the paths ourselves, or have to know anything about what platform we're on or the rules about paths.  

> Would it be awful to maintain this attempt small normalization? 

Basically, i have no idea how to review this function.  Is it correct?  I have no idea.  Is it complete?  I have no idea.  Will it throw and bring down something important?  I have no idea.  

> A lot of the file APIs that the runtime uses allow a prefixed / and so we prefix all of our paths with it to enable having the same path representation on Windows Mac and Linux.

Why can't URIs handle this?  How are URIs expected to work on Mac and Linux then?

NTaylorMullen replied:
> Why can't URIs handle this?

That would be the ideal solution; however, changing URIs is a much more difficult change to do than to ensure we normalize the input to a `Uri` construction. 

As for how URIs are expected to work on Mac/Linux: I'm not fully confident in the internals of Uri but filepaths being prefixed with `/` is the normal in Linux/Mac.

> Basically, i have no idea how to review this function. Is it correct? I have no idea. Is it complete? I have no idea. Will it throw and bring down something important? I have no idea.

I imagine you have a centralized Uri normalization flow when attempting to go from file path => Uri given users can put pretty much anything into a line pragma. 

This function's "correctness" and "completness" should be evaluated on its return value; aka, if it looks like a valid Uri then the Roslyn system shouldn't fall over even if the addressable location of the Uri cannot be found (that's the worse case).  

"Will it bring down something important" i'm making the statement of "if it looks like a valid Uri then Roslyn shouldn't fall over" because Uri's are sent to your language server from the LSP client and are outside of Roslyn's control in terms of what format they're sent. Aka, you already need to have resilience to the "wellformedness" of the Uri.

jasonmalinowski replied:
Yeah, this bit terrifies me as well, simply because it means this bit has to be special. Special is bad. Special leads to bugs. How do we keep a proper test suite to ensure that we don't screw this up, and this function works right?

The OS checking also worries me,simply because our LSP shouldn't depend on OS. If we had a unit test on this (which we should have no matter what....) you couldn't get this to pass on both sides without conditioning the unit test, right? And if LSP is being used cross-machine, what should be happening then?

NTaylorMullen replied:
I've filed an [issue](https://github.com/dotnet/aspnetcore/issues/19948) on our side to stop normalizing windows paths down to `/C:/path/to/something.razor`. Until we're able fix this issue (probably by end of April) would it be possible to take on the burden of this normalization?

CyrusNajmabadi replied:
I've very hestiant about that, unless we have confirmation that we can rip this code out at a certain date no matter what.  The main problem is that we have tons of cases of teams asking us to do this sort of thing, but then never following through on their end. (why would they?  the problem is now 'gone' because of the code in Roslyn).  So we have hacks and hacks that have to live forever.

My strong preference is that if this is going to be fixed on the razor side that we just fix that there.  Then we avoid the problem entirely.  What's the downside of just waiting on htat?  End of april is only 6 weeks away.  Is it necessary to have things fixed prior to that (and, if so, couldn't we just fix it faster in razor)?

NTaylorMullen replied:
> The main problem is that we have tons of cases of teams asking us to do this sort of thing, but then never following through on their end.

I totally understand! Some differences in our case though is that I personally have the decision power on what we do for which release and in the coming months we're going to be interacting a LOT. I personally will ensure this hack doesn't live forever.



> My strong preference is that if this is going to be fixed on the razor side that we just fix that there. Then we avoid the problem entirely. What's the downside of just waiting on htat? End of april is only 6 weeks away.

So sadly begin/mid April is when we're code completing for \\build in the attempt to deliver a quality "cloud" Razor experience. This item (delivering LSP diagnostics for Razor files) is one of our deliverables for that timeframe.

> couldn't we just fix it faster in razor?

Currently we're slammed with items attempting to get this \\Build scenario to light up and mucking with how our LanguageServer understands file paths in Razor is a significantly large item. We will absolutely do it but given its size we wont have time to do it prior to locking down for \\build.


--------------

So all that being said

> confirmation that we can rip this code out at a certain date no matter what.

To ensure we have enough time to complete the work and to account for "massiveness" on our end to fix it; can we compromise and say you all can 100% rip this out at the end of May (we will push for end of April though) and break the Razor experience? Oh and IF this issue causes any other issues you can rip it out immediately and we can wait for the Razor fix (losing Razor diagnostics of course, but don't want to cause a burden).

I've assigned the work item to our current and next iterations so it wont be lost and hell, even this convo makes it stick top-of-mind that we need to fix this 😄 .

If we can agree on an end-of-may no matter what rip-out date I'll update the issue on our end to reflect the agreement.

jasonmalinowski replied:
> To ensure we have enough time to complete the work and to account for "massiveness" on our end to fix it; can we compromise and say you all can 100% rip this out at the end of May (we will push for end of April though) and break the Razor experience?

Unfortunately "you can just break us" still turns into a "no we can't" once the reality of ship schedules and perhaps automated tests comes in. But the idea at least I can get behind.

Thanks for filing a tracking item.

CyrusNajmabadi replied:
> Currently we're slammed with items attempting to get this \Build scenario to light up and mucking with how our LanguageServer understands file paths in Razor is a significantly large item. We will absolutely do it but given its size we wont have time to do it prior to locking down for \build.

:-/

I really want to believe this to be true.  But we've been on the receiving end of those same promises so many times in the past.  We can run this up our end.  But my requirement would be that we are allowed to set a date to remove this, and there is no ability to pushback on that work actually happening.  And problems caused by removal would be entirely on Razor and not on Roslyn.  i.e. if we can't RI because of this, it would be on Razor to drop everything to make it possible.

That said, i don't think it would be possible for people on our end to approve even that.  i.e. no one is going to want to approve something that potentially impacts our own ability to meet ship dates :-/

NTaylorMullen replied:
> we are allowed to set a date to remove this, and there is no ability to pushback on that work actually happening

I'm absolutely with you here. Can we set May 31st as the absolute date, guaranteed no pushback from us?

> And problems caused by removal would be entirely on Razor and not on Roslyn. i.e. if we can't RI because of this, it would be on Razor to drop everything to make it possible.

Absolutely. In my original statement I even said, if this causes ANY other issues at all you can immediately remove it no questions asked.

> That said, i don't think it would be possible for people on our end to approve even that. i.e. no one is going to want to approve something that potentially impacts our own ability to meet ship dates :-/

If it'd make you feel better I'm willing to start that conversation with Jinu to get buy off from her as well. I see this as a super minor normalization that we intend to remove. Also, given that i'm the planner and executor on the Razor side I will make this happen.

NTaylorMullen replied:
@jinujoseph and I talked this over and after May 31st if we've (ASP.NET Core Razor) have yet to fix the file path normalization issue on our end you're more than welcome to remove it. Also, if this file path normalization breaks any of the other Roslyn related experiences you are also welcome to remove the logic immediately. We understand the removal would break Razor's LSP diagnostic reporting and are OK with that happening. The product would be alpha quality and at most would be broken for a few weeks; more importantly we've already funded the reaction on our side and have confidence this will get done long before that date.

/cc @mkArtakMSFT

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+                await _jsonRpc.NotifyWithParameterObjectAsync(Methods.TextDocumentPublishDiagnosticsName, publishDiagnosticsParams).ConfigureAwait(false);
+            });
+
+            await Task.WhenAll(publishTasks).ConfigureAwait(false);
```
do we have to be parallel?  This scares me *in proc*.  This can saturate cores, and cause lots of garbage to be created (causing gc pauses/hangs).  in-proc that can seriously degrade the editing experience.  We try to avoid this if possible.

dibarbet replied:
Nope - was just trying to be too smart for my own good.  Didn't know it would cause issues.  Will change to serialize it.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+            // These diagnostics will be reported with DiagnosticDataLocation.OriginalFilePath = generated C# file, and DiagnosticDataLocation.MappedFilePath = imported razor file.
+            // This means that in general we could have diagnostics produced by one generated file that map to many different actual razor files.
+            // We can't filter them out as we don't know which razor file(s) the underlying generated C# document actually maps to, and which are just imported.
+            // So we publish them all and let them get de-duped.
```
wrap this at more reasonable level (i.e. 100 chars).  i use the rewrap extension to make that better.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:
```diff
@@ -59,6 +60,22 @@ internal static class ProtocolConversions
+                filePath = filePath.Substring(1);
+            }
+
+            return new Uri(filePath, UriKind.Absolute);
```
can this throw for other reasons?

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+            await Task.WhenAll(publishTasks).ConfigureAwait(false);
         }
 
-        private async Task<LanguageServer.Protocol.Diagnostic[]> GetDiagnosticsAsync(Document document, CancellationToken cancellationToken)
```
`ImmutableArray<Diagnostic>`?

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
-                // Only the unnecessary diagnostic tag is currently supported via LSP.
-                Tags = diagnostic.CustomTags.Contains("Unnecessary") ? new DiagnosticTag[] { DiagnosticTag.Unnecessary } : Array.Empty<DiagnosticTag>()
-            }).ToArray();
+                var filePath = diagnosticData.DataLocation?.MappedFilePath ?? document.FilePath;
```
i hate this too.  1. Document.FilePath was never supposed to be used.    2. when does a diagnostic not have a data location??

dibarbet replied:
Not sure - but it's nullable.  Should I rather always use the original file path from the data location and just throw if null?

jasonmalinowski replied:
> when does a diagnostic not have a data location??

I could imagine diagnostics generally may not have a location: consider things like when the compiler hits an IO error and has to report that somehow. That said, I'd imagine all the diagnostics you get from the diagnostic service here *would* since they're tied to an ID.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+                    Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnosticData.Severity),
+                    Range = GetDiagnosticRange(diagnosticData.DataLocation, text),
+                    // Only the unnecessary diagnostic tag is currently supported via LSP.
+                    Tags = diagnosticData.CustomTags.Contains("Unnecessary") ? new DiagnosticTag[] { DiagnosticTag.Unnecessary } : Array.Empty<DiagnosticTag>()
```
1. should we not just pass along all tags?
2. WellKnownDiagnosticTags.Unnecessary

dibarbet replied:
LSP only supports the unnecessary tag.  we could pass the others, but I'd prefer to be explicit about what is passed rather than accidentally pass stuff through.

jasonmalinowski replied:
Yeah, the last thing we want to do is pass somebody's custom tag that breaks the LSP client. That said, Cyrus's WellKnownDiagnosticTags is good.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+            {
+                // This diagnostic is not restricted to a specific LSP client, so just pass it through
+                return true;
+            }
```
i don't get how the presence of this property indicates this.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+            }
+
+            // We only include this diagnostic if it directly matches our name.
+            return diagnosticClientName == _clientName;
```
this seems so fragile.

jasonmalinowski replied:
@CyrusNajmabadi I'm all ears for a better suggestion; this is the best we all could come up with...

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorLanguageClient.cs:
```diff
@@ -0,0 +1,118 @@
+    /// Defines the LSP server for Razor C#.  This is separate so that we can
+    /// activate this outside of a liveshare session and publish diagnostics
+    /// only for razor cs files.
+    /// TODO - This can be removed once C# is using LSP for diagnostics.
```
tracking issue please.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorLanguageClient.cs:
```diff
@@ -0,0 +1,118 @@
+        public Task<Connection> ActivateAsync(CancellationToken token)
+        {
+            var (clientStream, serverStream) = FullDuplexStream.CreatePair();
+            _ = new RazorLanguageServer(serverStream, serverStream, _languageServerProtocol, _workspace, _diagnosticService);
```
1. i hate that we just toss this on the floor.
2. what happens if ActivateAsync is called multiple times?

dibarbet replied:
1.  What should be done with it?  the lsp client just needs the stream connection
2.  It won't, guaranteed by the contract with the lsp client

CyrusNajmabadi replied:
1. Store as a field in the class. Tht way we can see it rooted if we get a dump :)
2. ok, can we throw then?

i.e. store this is a field.  if we get called again, throw that this is invalid.  I hate implicit contracts :)

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorLanguageServer.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Lsp
+{
+    class RazorLanguageServer : InProcLanguageServer
```
internal

jasonmalinowski replied:
```suggestion
    internal sealed class RazorLanguageServer : InProcLanguageServer
```

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorLanguageServer.cs:
```diff
@@ -0,0 +1,28 @@
+        protected override Task PublishDiagnosticsAsync(Document document)
+        {
+            // TODO - Filter razor diagnostics.
+            return base.PublishDiagnosticsAsync(document);
```
tracking issue?

jasonmalinowski replied:
Actually @dibarbet I wasn't sure what the intent was here then -- is the other filtering I implemented what you intended to go here?

dibarbet replied:
yes - I wasn't sure where it was going to happen.  Am going to remove this class entirely since we can just use the base.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseDiagnosticListTable.LiveTableDataSource.cs:
```diff
@@ -185,6 +189,12 @@ private static bool ShouldInclude(DiagnosticData diagnostic)
+                if (diagnostic.Properties.ContainsKey(nameof(DocumentPropertiesService.DiagnosticsLspClientName)))
+                {
+                    return false;
+                }
```
this seems awful.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/ServicesVSResources.resx:
```diff
@@ -1372,6 +1372,9 @@ I agree to all of the foregoing:</value>
     <value>Some color scheme colors are being overridden by changes made in the Environment &gt; Fonts and Colors options page. Choose `Use Defaults` in the Fonts and Colors page to revert all customizations.</value>
   </data>
+  <data name="Razor_CSharp_Language_Server_Client" xml:space="preserve">
+    <value>Razor C# Language Server Client</value>
```
why does this need a localized name?  when is this shown to the user?

dibarbet replied:
If an lsp command takes a long time, it pops up a dialog showing this

jasonmalinowski replied:
Should we pick a series of words that mean something to our customers then? (I'm not sure if this is a series of words that makes sense to *us*... :smile:)

dibarbet replied:
If you have suggestions, I'm all ears.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:
```diff
@@ -224,36 +225,48 @@ public async Task<Diagnostic> ToDiagnosticAsync(Project project, CancellationTok
+            var dataLocationStartLine = (useMapped ? dataLocation?.MappedStartLine : dataLocation?.OriginalStartLine) ?? 0;
+            var dataLocationStartColumn = (useMapped ? dataLocation?.MappedStartColumn : dataLocation?.OriginalStartColumn) ?? 0;
+            var dataLocationEndLine = (useMapped ? dataLocation?.MappedEndLine : dataLocation?.OriginalEndLine) ?? 0;
+            var dataLocationEndColumn = (useMapped ? dataLocation?.MappedEndColumn : dataLocation?.OriginalEndColumn) ?? 0;
```
when is useMapped false?

dibarbet replied:
The tagger doesn't use the mapped spans - http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Diagnostics/DiagnosticData.cs,828891a1bfd06025

http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs,240

I didn't want to change that so added the flag.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:
```diff
@@ -224,36 +225,48 @@ public async Task<Diagnostic> ToDiagnosticAsync(Project project, CancellationTok
+            var dataLocationEndLine = (useMapped ? dataLocation?.MappedEndLine : dataLocation?.OriginalEndLine) ?? 0;
+            var dataLocationEndColumn = (useMapped ? dataLocation?.MappedEndColumn : dataLocation?.OriginalEndColumn) ?? 0;
+
+            if (dataLocationStartLine >= lines.Count)
```
how could this possibly happen?

dibarbet replied:
Not sure - just moved the logic.  theoretically, if the diagnostic was published for a different version of the text, it could be outside the bounds?

CyrusNajmabadi replied:
>  theoretically, if the diagnostic was published for a different version of the text,

When does that happen?  In roslyn, everything should be snapshotted.  So it shouldn't be possible to have diagnostics published for different versions right?  Does LSP change that?

dibarbet replied:
LSP doesn't change that, just going off of this comment - http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs,202

CyrusNajmabadi replied:
:( :( :( :( :( :( :( :( :( :( 

Ok... :(

I guess we have to be safe from this case then.

jasonmalinowski replied:
Do we have a tracking bug for that cleanup though? The internal bug in that code regarding the NFW is two and a half years old...

dibarbet replied:
I think the internal bug is referring to something else.  The link I posted was to the comment that was expecting the diagnostic to be published for a different document version, and the corresponding mapping code to ensure it's in bounds for the current version.

CyrusNajmabadi replied:
> The link I posted was to the comment that was expecting the diagnostic to be published for a different document version, and the corresponding mapping code to ensure it's in bounds for the current version.

It's super weird though... so we have a diagnostic for a span that seems to no longer exist... but we're trying to map that span to our current doc.  

My pref would be to just drop the diagnostic on the floor then.  

--

However, if htis is what we already do, i guess i'm ok with it.  I just don't like that the actual design here is just "we have snapshots.  we don't use them... but we have them.  and here's some data about some random version that you have to reconcile with whatever actual version you have".

It's basically broken by-design and doesn't follow *any* of our other snapshot-oriented systems we have in roslyn.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:
```diff
@@ -336,13 +349,28 @@ public static DiagnosticData Create(Diagnostic diagnostic, Document document)
 
+            var additionalProperties = GetAdditionalProperties(document, diagnostic);
+
+            var documentPropertiesService = document.Services.GetService<DocumentPropertiesService>();
```
we have a service that isn't an interface?

jasonmalinowski replied:
Yes.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:
```diff
@@ -336,13 +349,28 @@ public static DiagnosticData Create(Diagnostic diagnostic, Document document)
                 location,
                 additionalLocations,
-                GetAdditionalProperties(document, diagnostic));
+                additionalProperties);
```
i'm confused why this needs to be in the diagnostic.  if we can already ask the Document for it's `DiagnosticsLspClientName`, then why can't people just look at that instead of the diagnostic itself?

jasonmalinowski replied:
So the DiagnosticData doesn't hold onto the Document or DocumentId as far as I can tell; so if we didn't capture the value here it's not easy for other layers (like the error list) to re-filter -- once you're past this point you don't know which document it came from. And since file names can be remapped, that would make it hard to re-find the document as best I can tell.

It's possible I'm missing something else here but it seemed like GetAdditionalProperties was "where" we did the last chance ability to capture from the Document because beyond here it was "too late".

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/Host/DocumentService/DocumentPropertiesService.cs:
```diff
@@ -16,5 +18,11 @@ internal class DocumentPropertiesService : IDocumentService
+
+        /// <summary>
+        /// The LSP client name that should get the diagnostics produced by this document; any other source
+        /// will not show these diagnostics. If null, the diagnostics do not have this special handling.
```
can you expand on this (a lot).  what do you mean by "any other source will not show these diagnostics"?  who will show the diagnostics then?

jasonmalinowski replied:
The error list would be the thing that this is primarly being excluded from. The idea is the Razor folks can query this and they're responsible for final display.

@dibarbet Do you want me to reword this in your PR or are you good?

dibarbet replied:
I should be able to

jasonmalinowski commented:
> i hate changes like this where we basically backchannel data to try to couple different components together tightly to get them to change behavior.

@CyrusNajmabadi Suggestions? I don't like what we're doing either, but we couldn't come up with a better idea.

jasonmalinowski commented on src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:
```diff
@@ -59,6 +60,22 @@ internal static class ProtocolConversions
+            }
+
+            // Remove preceding slash if we're on Window as it's an invalid URI.
+            if (filePath.StartsWith("/") && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
To clarify, so what is the paths we're getting from Razor? It's like /C:/foo/bar/baz.txt? Or something else?

NTaylorMullen replied:
You got it exactly!

jasonmalinowski commented on src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:
```diff
@@ -59,6 +60,22 @@ internal static class ProtocolConversions
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            // Remove preceding slash if we're on Window as it's an invalid URI.
```
Let's point this to https://github.com/dotnet/aspnetcore/issues/19948 now that the bug has been filed.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+                {
+                    Line = startLine.Line,
+                    Character = startLine.Character,
+                },
```
Isn't this ProtocolConversions.LinePositionToPosition? Or even better we don't have an API that just returns a LinePositionSpan?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorLanguageClient.cs:
```diff
@@ -0,0 +1,118 @@
+        /// Signals that the extension has been loaded.  The server can be started immediately, or wait for user action to start.  
+        /// To start the server, invoke the <see cref="StartAsync"/> event;
+        /// </summary>
+        public async Task OnLoadedAsync()
```
Did we want to have a base class for this and our regular LanguageClient so all of this can be reused? The base class would be abstract as a few things need to change, but at least it's a relatively small bit then.

jasonmalinowski commented on src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:
```diff
@@ -224,36 +225,48 @@ public async Task<Diagnostic> ToDiagnosticAsync(Project project, CancellationTok
             var endLinePosition = new LinePosition(endLine, endColumn);
             SwapIfNeeded(ref startLinePosition, ref endLinePosition);
 
+            return (startLinePosition, endLinePosition);
```
This is what LinePositionSpan is for.

jasonmalinowski commented on src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:
```diff
@@ -224,36 +225,48 @@ public async Task<Diagnostic> ToDiagnosticAsync(Project project, CancellationTok
+
+        public static TextSpan GetTextSpan(DiagnosticDataLocation? dataLocation, SourceText text)
+        {
+            (var startLinePosition, var endLinePosition) = GetLinePositions(dataLocation, text, useMapped: false);
```
If this returns a LinePositionSpan then you can simplify again.

NTaylorMullen commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/AbstractLiveShareLanguageServerClient.cs:
```diff
@@ -0,0 +1,104 @@
+            Contract.ThrowIfFalse(_languageServer == null, "This language server has already been initialized");
+
+            var (clientStream, serverStream) = FullDuplexStream.CreatePair();
+            _languageServer = new InProcLanguageServer(serverStream, serverStream, _languageServerProtocol, _workspace, _diagnosticService, clientName: null);
```
This shouldn't be passing `clientName` `null` should it?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +226,163 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+
+        private static V GetOrValue<K, V>(Dictionary<K, V> dict, K key, V defaultValue)
+        {
+            if (dict.TryGetValue(key, out var immutableArray))
```
```suggestion
            if (dict.TryGetValue(key, out var value))
```

Field name doesn't make sense otherwise. Do you also want to add it this to https://github.com/dotnet/roslyn/blob/master/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/IDictionaryExtensions.cs instead?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +226,163 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
-        private async Task PublishDiagnosticsAsync(Document document)
+        /// <summary>
+        /// Stores the last published LSP diagnostics with the Roslyn document that they came from.
+        /// This is useful in the following scenario.  Imagine we have documentA which has contributions from mapped files m1 and m2.
```
(this is mostly a terminology question since I guess I imagine it the other way. Phrasing applies to rest of the comment.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +226,163 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
             var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
 
-            return diagnostics.Select(diagnostic => new LanguageServer.Protocol.Diagnostic
+            // Razor documents can import other razor documents
+            // https://docs.microsoft.com/en-us/aspnet/core/mvc/views/layout?view=aspnetcore-3.1#importing-shared-directives
+            // https://docs.microsoft.com/en-us/aspnet/core/blazor/layouts?view=aspnetcore-3.1#centralized-layout-selection
+            // The imported files contents are added to the content of the generated C# file, so we get diagnostics
+            // for both the c# contents in the original razor document and for any of the content in any of the imported
+            // files when we query diagnostics for the generated C# file.
+            // These diagnostics will be reported with DiagnosticDataLocation.OriginalFilePath = generated C# file and
+            // DiagnosticDataLocation.MappedFilePath = imported razor file.
+            // This means that in general we could have diagnostics produced by one generated file that map to many different actual razor files.
+            // We can't filter them out as we don't know which razor file(s) the underlying generated C# document actually maps to
+            // and which are just imported.  So we publish them all and let them get de-duped.
```
I admit I'm not entirely following what this whole comment is getting at. It implies there's some trickery coming and then the ultimate stuff isn't really trickery at all?

dibarbet replied:
It's an old comment I can probably delete / steal parts of and add to other comments.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +226,163 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+        /// </summary>
+        private readonly Dictionary<DocumentId, ImmutableHashSet<Uri>> _documentsToPublishedUris = new Dictionary<DocumentId, ImmutableHashSet<Uri>>();
+
+        internal async Task PublishDiagnosticsAsync(Document document)
```
Is this serialized so we don't have to worry about parallel operations?

dibarbet replied:
Yes, and so that we publish diagnostics 'in order' - since diagnostics are replaced we don't want to publish old diagnostics after new diagnostics.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/AbstractLiveShareLanguageServerClient.cs:
```diff
@@ -0,0 +1,107 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
+{
+    internal abstract class AbstractLiveShareLanguageServerClient : ILanguageClient
```
Why "LiveShare" in the name? Because it's used for other stuff maybe too?

dibarbet replied:
Right - can rename now that it's being used by razor as well.

jasonmalinowski commented on src/VisualStudio/LiveShare/Impl/FindAllReferencesHandler.cs:
```diff
@@ -55,6 +55,7 @@ public async Task<object[]> HandleAsync(LSP.ReferenceParams request, RequestCont
 
             await findUsagesService.FindReferencesAsync(document, position, context).ConfigureAwait(false);
 
+
```
Extra blank line.

jasonmalinowski commented on src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:
```diff
@@ -0,0 +1,470 @@
+            /// </summary>
+            private int _currentNumberOfCallbacks;
+
+
```
Extra blank line.

jasonmalinowski commented on src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:
```diff
@@ -0,0 +1,470 @@
+            [JsonRpcMethod(LSP.Methods.TextDocumentPublishDiagnosticsName)]
+            public Task OnDiagnosticsPublished(JToken input)
+            {
+                lock(_lock)
```
```suggestion
                lock (_lock)
```

jasonmalinowski commented on src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:
```diff
@@ -0,0 +1,470 @@
+            // 1.  from doc1 with id.
+            var results = await RunPublishDiagnosticsAsync(workspace, diagnosticsMock.Object, 1, document).ConfigureAwait(false);
+
+            // We expect one publish notification for the diagnostic on the document.
```
Consider `var result = Assert.Single(results)` once which both removes all the .Single()s everywhere else, but also gives a nice assert failure if it isn't right.

jasonmalinowski commented on src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:
```diff
@@ -0,0 +1,470 @@
+            var results = await RunPublishDiagnosticsAsync(workspace, diagnosticsMock.Object, expectedNumberOfCallbacks: 2, document).ConfigureAwait(false);
+
+            Assert.Equal(2, results.Count);
+            Assert.Equal("id1", results.Single(p => p.Uri == new Uri(document.FilePath + "m1")).Diagnostics.Single().Code);
```
Work having this return a dictionary or something so you don't have to write the whole `results.Single(p => p.Uri == new Uri(` thing multiple times?

dibarbet replied:
I actually started with dictionary, but it was a bit of a struggle.  With a dictionary, I can only verify the end result (after all publish notifications are complete) since we'd be replacing diagnostics for a particular URI as we publish.  I found it much easier to write tests (and I think more correct) when I can verify that each incremental publish notification produces the correct output.

jasonmalinowski commented on src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:
```diff
@@ -0,0 +1,470 @@
+
+            Assert.Equal(expectedUri, results[1].Uri);
+            Assert.Equal(2, results[1].Diagnostics.Length);
+            Assert.True(results[1].Diagnostics.Contains(d => d.Code == "doc1Diagnostic"));
```
There is an Assert.Contains() that takes a lambda you can pass; that way if this fails you don't just get "False".

jasonmalinowski commented on src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:
```diff
@@ -0,0 +1,470 @@
+            SetupMockWithDiagnostics(diagnosticsMock, documents[0].Id, documentOneDiagnostic);
+            SetupMockWithDiagnostics(diagnosticsMock, documents[1].Id, documentTwoDiagnostic);
+
+            // Publish two document change diagnostic notifications ->
+            // 1.  doc1 with doc1Diagnostic = mapped file m1.
+            // 2.  doc2 with doc2Diagnostic = mapped file m1.
+            //
+            // We expect two publish diagnostic notifications ->
+            // 1.  from m1 with doc1Diagnostic (from 1 above).
+            // 2.  from m1 with doc1Diagnostic and doc2Diagnostic (from 2 above adding doc2Diagnostic to m1).
```
Loving the comments by the way this is very clear.

jasonmalinowski commented on src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:
```diff
@@ -0,0 +1,470 @@
+
+            Assert.Equal(expectedUri, results[3].Uri);
+            Assert.True(results[3].Diagnostics.IsEmpty());
+        }
```
This probably needs some asserts here that the underling maps are empty otherwise they're potentially leaking...

dibarbet commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +226,163 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+                {
+                    // We published an empty set of diagnostics for this uri.  We no longer need to keep track of this mapping
+                    // since there will be no previous diagnostics that we need to clear out.
+                    _documentsToPublishedUris[document.Id] = _documentsToPublishedUris[document.Id].Remove(fileUri);
```
hmm yes I think if we remove diagnostics from all fileUris that are part of this documentId we cab remove the document.Id entry as well.

Can probably just do a check outside the loop to see if the key has any values and remove it if not.  Will do.

dibarbet commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -219,27 +226,163 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+                else
+                {
+                    // There are no new diagnostics for this document and file uri, if we're tracking it we can stop.
+                    documentsToPublishedDiagnostics.Remove(document.Id);
```
fixed!

dibarbet commented:
@jasonmalinowski I believe I addressed the leaks issue.  I haven't addressed the bit about the nested dictionary - would prefer to do that in a followup PR so I can at least unblock @NTaylorMullen 

Also fixed flakiness in the tests due to hashset not returning a consistent ordering (forgot about that... 🤦‍♂ ).

jasonmalinowski commented on src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:
```diff
@@ -131,14 +134,17 @@ public async Task RemoveDiagnosticTestAsync()
             Assert.Equal(new Uri(document.FilePath), results[1].Uri);
             Assert.True(results[1].Diagnostics.IsEmpty());
+
+            Assert.False(languageServer._documentsToPublishedUris.Keys.Any());
```
(ditto for other uses)

dibarbet commented:
Merging to unblock @NTaylorMullen - will address the rest of the feedback in a followup this afternoon.
